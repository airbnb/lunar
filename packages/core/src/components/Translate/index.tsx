import React from 'react';
import Core from '../..';
import { TranslateParams, TranslateProps } from '../../types';

export type Props = TranslateProps;

/** Translate a phrase with a context and dynamic params. */
export default class Translate extends React.PureComponent<Props> {
  static defaultProps = {
    html: false,
  };

  static phrase(phrase: string, params: TranslateParams, context: string): string {
    return Core.translate(phrase, params, context);
  }

  render() {
    const { translatorComponent: Translator } = Core.settings;
    const { children, phrase, context, ...params } = this.props;

    if (!Translator) {
      return <span>{Core.translate(phrase, params as any, context)}</span>;
    }

    return <Translator phrase={phrase} context={context} {...(params as any)} />;
  }
}
