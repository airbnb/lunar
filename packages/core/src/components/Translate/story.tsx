import React from 'react';
import Text from '../Text';
import Translate from '.';

export default {
  title: 'Core/Translate',
  parameters: {
    inspectComponents: [Translate],
  },
};

export function displayAMessageWithAnEditorRelatedContext() {
  return (
    <Text>
      <Translate
        phrase="This content should be translated."
        context="This message is for translation editors."
      />
    </Text>
  );
}

displayAMessageWithAnEditorRelatedContext.story = {
  name: 'Display a message with an editor related context.',
};

export function interpolateVariablesAlsoSupportHtml() {
  return (
    <Text>
      <Translate
        html
        phrase="Hello %{name}!"
        name={<b>Bruce</b>}
        context="This message is for translation editors."
      />
    </Text>
  );
}

interpolateVariablesAlsoSupportHtml.story = {
  name: 'Interpolate variables. Also support HTML.',
};

export function handleContextualMessagesBasedOnCounts() {
  return (
    <>
      <Text>
        <Translate
          phrase="%{smartCount} item||||%{smartCount} items"
          smartCount={0}
          context="This message is for translation editors."
        />
        <br />
        <br />
        <Translate
          phrase="%{smartCount} item||||%{smartCount} items"
          smartCount={1}
          context="This message is for translation editors."
        />
        <br />
        <br />
        <Translate
          phrase="%{smartCount} item||||%{smartCount} items"
          smartCount={2}
          context="This message is for translation editors."
        />
      </Text>
    </>
  );
}

handleContextualMessagesBasedOnCounts.story = {
  name: 'Handle contextual messages based on counts.',
};

export function returnAStringInsteadOfRenderingAComponent() {
  return (
    <div>
      {Translate.phrase(
        'Hello %{name}',
        { name: 'Bruce' },
        'This message is for translation editors.',
      )}
    </div>
  );
}

returnAStringInsteadOfRenderingAComponent.story = {
  name: 'Return a string instead of rendering a component.',
};
