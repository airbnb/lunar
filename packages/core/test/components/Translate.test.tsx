import React from 'react';
import { shallow } from 'enzyme';
import Core, { Settings } from '../../src';
import Translate from '../../src/components/Translate';

describe('<Translate />', () => {
  let oldSettings: Required<Settings>;

  function T() {
    return null;
  }

  beforeEach(() => {
    oldSettings = Core.settings;

    Core.settings.translator = jest.fn();
    Core.settings.translatorComponent = T;
  });

  afterEach(() => {
    Core.settings = oldSettings;
  });

  it('renders a translator component', () => {
    const wrapper = shallow(<Translate phrase="Foo" bar={123} context="Baz" />);

    expect(wrapper.find(T)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with a `k` prop', () => {
    const wrapper = shallow(<Translate k="foo" phrase="Foo" bar={123} context="Baz" />);

    expect(wrapper.find(T).prop('k')).toBe('foo');
  });

  it('renders without translator component', () => {
    Core.settings.translatorComponent = null;

    const wrapper = shallow(<Translate phrase="Foo" bar={123} context="Baz" />);

    expect(wrapper.find(T)).toHaveLength(0);
    expect(wrapper).toMatchSnapshot();
  });

  describe('phrase()', () => {
    it('passes params to translator', () => {
      Translate.phrase('Foo', { bar: 123 }, 'Baz');

      expect(Core.settings.translator).toHaveBeenCalledWith('Foo', { bar: 123 }, 'Baz');
    });

    it('supports object options', () => {
      Translate.phrase('Foo', { bar: 123 }, { context: 'Baz', key: 'foo' });

      expect(Core.settings.translator).toHaveBeenCalledWith(
        'Foo',
        { bar: 123 },
        { context: 'Baz', key: 'foo' },
      );
    });
  });
});
