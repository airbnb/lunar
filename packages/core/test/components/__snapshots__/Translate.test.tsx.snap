// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Translate /> renders a translator component 1`] = `
<T
  bar={123}
  k="foo"
  phrase="Foo"
/>
`;

exports[`<Translate /> renders without translator component 1`] = `<span />`;
